pre-sync/oci-image-verification/main.go

```diff
@@ -0,0 +1,145 @@
+	if err != nil {
+		return fmt.Errorf("Error fetching access token: %v\n", err)
+	}
+	// Convert the output to a string and trim any trailing newline
+	accessToken := string(accessTokenBytes)
+	accessToken = accessToken[:len(accessToken)-1] // Remove the trailing newline
+
+	cmd := exec.Command("cosign", "login", "us-central1-docker.pkg.dev", "-u", "oauth2accesstoken", "-p", accessToken)
+	log.Printf("command %s", cmd.String())
+	output, err := cmd.CombinedOutput()
+	if err != nil {
+		return fmt.Errorf("gcloud auth ar failed: %s, output: %s", err, string(output))
+	}
+	log.Printf("result: %s, authorization done", string(output))
+	authorized = true
+	return nil
+}
+
+func handleWebhook(w http.ResponseWriter, r *http.Request) {
+
+	// Read the body
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		log.Printf("Failed to read request body: %v", err)
+		http.Error(w, "Failed to read request body", http.StatusBadRequest)
+		return
+	}
+
+	var admissionReview admissionv1.AdmissionReview
+	if err := json.Unmarshal(body, &admissionReview); err != nil {
+		log.Printf("Failed to unmarshal admission review: %v", err)
+		http.Error(w, "Failed to unmarshal admission review", http.StatusBadRequest)
+		return
+	}
+
+	// Extract old and new annotations
+	oldAnnotations, err := getAnnotations(admissionReview.Request.OldObject.Raw)
+	if err != nil {
+		log.Printf("Failed to extract old annotations: %v", err)
+	}
+
+	newAnnotations, err := getAnnotations(admissionReview.Request.Object.Raw)
+	if err != nil {
+		log.Printf("Failed to extract new annotations: %v", err)
+	}
+
+	// Log old and new annotations for comparison
+	log.Printf("Old Annotations: %v", oldAnnotations)
+	log.Printf("New Annotations: %v", newAnnotations)
+
+	// Send the admission response
+	response := &admissionv1.AdmissionResponse{
+		UID: admissionReview.Request.UID,
+	}
+
+	// Compare the annotations and check for changes
+	if newAnnotations[SourceURL] != oldAnnotations[SourceURL] ||
+		newAnnotations[SourceCommit] != oldAnnotations[SourceCommit] {
+		log.Printf("Detected annotation changes")
+		auth()
+		// Validate image using cosign
+		if err := validateImage(newAnnotations[SourceURL]); err != nil {
+			log.Printf("Image validation failed: %v", err)
+			response.Allowed = false
+			response.Result = &metav1.Status{
+				Message: fmt.Sprintf("Image validation failed: %v", err),
+			}
+		} else {
+			log.Printf("Image validation successful for %s", newAnnotations[SourceURL])
+			response.Allowed = true
+		}
+	} else {
+		log.Printf("No annotation changes detected")
+	}
+
+	admissionReview.Response = response
+	if err := json.NewEncoder(w).Encode(admissionReview); err != nil {
+		log.Printf("Failed to encode admission response: %v", err)
+	}
+}
+
+func main() {
+	http.HandleFunc("/validate", handleWebhook)
+	log.Println("Starting webhook server on port 8443...")
+	log.Fatal(http.ListenAndServeTLS(":8443", "/tls/tls.crt", "/tls/tls.key", nil))
```

Comment:
Config Sync's admission webhook uses `10250` to avoid adding a firewall rule for a private cluster. We may want to use the same port.

https://github.com/GoogleContainerTools/kpt-config-sync/blob/main/pkg/webhook/configuration/constants.go#L41

Reply:
Done